components/script/dom/htmlinputelement.rs

```diff
@@ -886,6 +873,23 @@ impl HTMLInputElement {
             _ => ()
         }
     }
+
+    fn selection_direction(&self) -> SelectionDirection {
+        self.textinput.borrow().selection_direction
+    }
+
+    fn set_selection_range(&self, start: u32, end: u32, direction: SelectionDirection) {
+        self.textinput.borrow_mut().selection_direction = direction;
+        self.textinput.borrow_mut().set_selection_range(start, end);
+        let window = window_from_node(self);
+        let _ = window.user_interaction_task_source().queue_event(
```

Comment:
Let's also add step annotations for this function. For example, this particular line here follows step 6 of the linked algorithm (albeit incorrectly).